aten/src/ATen/native/ComplexHelper.cpp

```diff
@@ -0,0 +1,54 @@
+#include <ATen/native/ComplexHelper.h>
+
+#include <ATen/native/TensorIterator.h>
+
+namespace at {
+namespace native {
+
+DEFINE_DISPATCH(complex_stub);
+DEFINE_DISPATCH(complex_polar_stub);
+
+Tensor& complex_out(Tensor& result, const Tensor& real, const Tensor& imag) {
+  TORCH_CHECK(real.scalar_type() == imag.scalar_type(),
+              "Expected object of scalar type ", real.scalar_type(), " but got "
+              "scalar type ", imag.scalar_type(), " for argument 'imag'");
+  // Sort of hacky, but necessary so that input dtypes don't get promoted.
+  auto iter = TensorIterator::comparison_op(result, real, imag,
```

Comment:
what is this for?

Reply:
ahh I see to do all the required iterator stuff

Reply:
what if `real` and `imag` were complex tensors? that shouldn't be allowed

Reply:
I think because the kernel is only dispatched for floating types it'll say e.g. `"complex_cpu" not implemented for 'ComplexFloat'`. This is the same behavior as if I called `torch.cos(x)` where `x` was an integer tensor.